csrc/scheduler/normalization_inner_outer.cpp

```diff
@@ -238,33 +240,53 @@ PersistentBufferStorageParams getPersistentBufferStorageParams(
   int64_t available_regs = scheduler_utils::register_file_size_56k;
   buffer_params.smem_overhead = smem_overhead;
 
-  // (1) init the buffer_params by putting all the persistent tensors in
-  // registers
-  buffer_params.regs_buffer_size = total_buffer_size;
-  buffer_params.smem_buffer_size = 0;
+  // (1) Try to put all the buffers into registers.
+  if (total_buffer_size <= available_regs) {
+    buffer_params.regs_buffer_size = total_buffer_size;
+    buffer_params.smem_buffer_size = 0;
+    buffer_params.has_enough_regs_and_smem = true;
+    return buffer_params;
+  }
 
-  // (2) If the available register is larger than current register buffer size,
-  // no need to move buffers to shared memory, return early.
-  if (buffer_params.regs_buffer_size <= available_regs) {
+  // (2) Consider using both register and shared memory.
+  // Start with all the cached input buffers in shared memory, they are loaded
+  // from global memory uses async copy which bypasses L1 cache. Outer reduction
+  // buffers are used to accumulate partial results of the outer reduction. They
+  // are not loaded from global memory and requires frequent read/write. So,
+  // they are always stored in registers.
+  // TODO: We may also move outer reduction buffers to shared
+  // memory to avoid segmentation when there are many outer reductions and
+  // hardware has larger shared memory, but these applications are rare and not
+  // considered here.
+  buffer_params.smem_buffer_size = cached_input_buffer_size;
+  buffer_params.regs_buffer_size = outer_reduction_buffer_size;
+  if (buffer_params.regs_buffer_size <= available_regs &&
+      buffer_params.smem_buffer_size <= available_smem) {
+    buffer_params.smem_persistent_buffers = buffer_params.project_to_input
+        ? persistent_buffer_info.projectable_buffer_inputs
+        : persistent_buffer_info.persistent_buffers;
     buffer_params.has_enough_regs_and_smem = true;
     return buffer_params;
   }
 
-  // (3) Relocate buffers to shared memory until the buffer size in registers is
-  // within the allowable limit.
+  // Moving outer reduction buffer to shared memory is not considered yet,
+  // set to false if the outer reduction buffer size exceeds the register size.
+  if (buffer_params.regs_buffer_size > available_regs) {
+    buffer_params.has_enough_regs_and_smem = false;
+    return buffer_params;
+  }
```

Comment:
This `if` branch will be removed in #3221. It should be removed with a slight change of the heuristics to avoid large regressions.